### PR TITLE
Allow to change default session with with

### PR DESCRIPTION
**Pull request by urbanmatthias**
This will change the default session in a `with` statement:

```
with SomeWrapperSession() as session:
    # default session will be session
```

- I guess this makes the requirement that all wrapper object must be instantiated with an explicitly provided session obsolete
- This will sometimes break behavior as you can see from the tests. I hope this happens only in artificial cases like the tests

### CODE REVIEW

File: osp/core/session/session.py [hunk 1]
```diff
@@ -18,13 +18,19 @@ def __init__(self):
         self._registry = Registry()
         self.root = None
         self.graph = rdflib.Graph()
+        self._previous_session = None
 
     def __enter__(self):
         """Establish the connection to the backend."""
+        from osp.core.cuds import Cuds
```
Comment (pablo-de-andres) [inline osp/core/session/session.py:25]: What would happen if you had nested `with` statements?
Wouldn't this always set the previous to `CoreSession`?

Comment (urbanmatthias) [reply osp/core/session/session.py:25]: You're right. The previous session should be a member of Cuds I guess

Comment (urbanmatthias) [reply osp/core/session/session.py:25]: I guess unittests should be added to this

**Comment by urbanmatthias** [comment]
The benefit that is already implemented:

with SomeSession() as session:
    w = ns.Wrapper(session=session)
    c = ns.City()  # this would previously be created in the core session
    w.add(c)  # this would previously copy the city
    
But the user is not interested in having objects in the core session. After the change, the City would have been created in the SomeSession

File: osp/core/session/session.py (hunk 1 above)
Comment (urbanmatthias) [reply osp/core/session/session.py:25]: okay you're not right

**Comment by kysrpex** [comment]
> The benefit that is already implemented:
> 
> ```
> with SomeSession() as session:
>     w = ns.Wrapper(session=session)
>     c = ns.City()  # this would previously be created in the core session
>     w.add(c)  # this would previously copy the city
> ```
> 
> But the user is not interested in having objects in the core session. After the change, the City would have been created in the SomeSession

Ok, then it's both a performance benefit (avoiding unnecessary deepcopy) and an usability benefit (the user can edit `c` directly  and see the changes reflected). Great!

File: osp/core/session/session.py (hunk 1 above)
Comment (kysrpex) [reply osp/core/session/session.py:25]: In any case, we need an unit test for this. Maybe I can write it , that would make the code review more efective.

**Comment by yoavnash** [comment]
Did you check the scenarios:
- Multiple depth-levels of 'with' statements
- Multiple 'with' statements inside one 'with' statement
?

**Comment by urbanmatthias** [comment]
There is no test yet, but it should work in these scenarios

File: osp/core/session/session.py (hunk 1 above)
Comment (kysrpex) [inline osp/core/session/session.py:25]: It makes not difference in practice, but would it not make sense to just import Cuds once at the top?

**Review by kysrpex — APPROVED** [review]
I think it's fine and seems to work in the tests.

**Review by yoavnash — APPROVED** [review]

**Comment by urbanmatthias** [comment]
@pablo-de-andres I merged. We should update the docs. I'll create an issue there

File: osp/core/session/session.py (hunk 1 above)
Comment (urbanmatthias) [reply osp/core/session/session.py:25]: Unfortunately, it sometimes happens that there are circular imports. I tried to put it at the top, got such an error and moved it there. Ideally, we should structure our imports s.t. no circular imports occur. I didn't check why it occured in this case.

**Pull request merged**